File: spoj/PT07Y/PT07Y.rb
class Graph

  def initialize(adj)
    @adj = adj
    @visited = []
  end

  def is_tree?(node=0, prev=nil)
    @visited << node
    adjacent(node, prev).each do |a|
      return false if @visited.include? a
      return false unless is_tree?(a, node)
    end

    if node == 0 && @visited.size < @adj.size
      return false
    end

    return true
  end

  private
    def adjacent(node, prev)
      adj = []
      @adj[node].each.with_index do |v, i|
        adj << i if (v && i != prev && i != node)
      end
      return adj
    end
end

if __FILE__ == $0
  N, M = STDIN.gets.strip.split.map { |val| val.to_i }
  adj = Array.new(N+1) { Array.new(N+1, false) }
  (1..M).each do |i|
    p, q = STDIN.gets.strip.split.map { |val| val.to_i }
    adj[p][q] = true
  end

  graph = Graph.new(adj)
  puts graph.is_tree? ? "YES" : "NO"
end

File: spoj/PT07Y/PT07Y.cpp
#include<cstdio>
#include<iostream>
#include<algorithm>
#include<cstdlib>
#include<cmath>
#include<cassert>
#include<vector>
#include<map>
#include<cstring>
#include<cassert>
#include<queue>

#define FOR(i,n) for(int i=0;i<n;i++)
#define FORS(i,a,n) for(int i=a;i<n;i++)
#define ERR(x) cerr<<#x<<" "<<x<<endl
#define pb push_back
#define FILL(a,b) memset(a,b,sizeof(a))
using namespace std;

typedef vector<int> VI;
typedef long long ll;
typedef long double ld;

inline int print(int d) { return printf("%d",d);}
inline int read(int &d) { return scanf("%d",&d);}

int N,M,tmp;
vector<int> graph[10001];
vector<int> visited;

bool isTree(int node, int prev) {
  if(std::find(visited.begin(), visited.end(), node) != visited.end()) {
    return false;
  }
  visited.push_back(node);
  for(std::vector<int>::size_type i = 0; i != graph[node].size(); i++) {
    if(graph[node][i] == node)
      continue;

    if(!isTree(graph[node][i], node))
      return false;
  }

  return true;
}

int main(int argc, char** args)
{
  read(N);
  read(M);

  int p, q;
  FOR(i,M) {
    read(p);
    read(q);
    graph[p].push_back(q);
  }

  if(isTree(1, 0))
    if(visited.size() == N)
      printf("YES");
    else
      printf("NO");
  else
    printf("NO");
}
